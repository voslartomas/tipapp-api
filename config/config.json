{
  "development": {
    "username": "root",
    "password": "",
    "database": "neymar.sqlite",
    "host": "localhost",
    "storage": "../neymar.sqlite",
    "dialect": "sqlite"
  },
  "test": {
    "username": "root",
    "password": null,
    "database": "database_test",
    "host": "127.0.0.1",
    "dialect": "mysql"
  },
  "production": {
    "username": "tvoslar",
    "password": "tipapptipapp",
    "database": "tipapp",
    "host": "tipapp.c2fgvnquam0j.eu-central-1.rds.amazonaws.com",
    "dialect": "postgres"
  }
}
