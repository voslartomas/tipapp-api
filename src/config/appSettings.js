module.exports.Settings = {
    db: {
        //path: 'C:\Users\talas\Desktop\bettingLeague\db\neymar.sqlite',
        path: './db/neymar.sqlite',
        databaseName: 'neymar.sqlite',
        username: '',
        password: '',
        host: 'localhost',
        port: 3000,
        dialect: 'sqlite',
        logging: console.log,
        typeValidation: true,
        pool: {
            max: 5,
            min: 0,
            idle: 10000
        },
        //storage: 'C:\Users\talas\Desktop\bettingLeague\db\neymar.sqlite'
    },
    connection: {
        port: 3000
    },
    initializationDone: false
}