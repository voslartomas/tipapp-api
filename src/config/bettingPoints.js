module.exports.bettingPoints = {
  hockey: {
    'matchWinner': 4,
    'exactScore': 8,
    'goalsOfOneTeam': 1,
    'goalsDifference': 2,
    'scorer': 6,
    'bestScorer': 2,
    'nhlSeries': 8,
    'nhlSeriesWithGames': 14,
    'nhlConference': 20,
    'nhlWinner': 40,
    'nhlMvp': 30,
    'wcGroup': 20,
    'wcLast': 20,
    'wcWinner': 40,
    'wcScorer': 30,
    'wcGoalkeeper': 30,
    'wcMvp': 30,
    'wcExactTotalGoals': 50,
    'wcTotalGoals': 20
  },
  football: {
    'matchWinner': 3,
    'matchWinnerWithDifference': 4,
    'draw': 4,
    'exactScore': 6,
    'scorer': 5,
    'bestScorer': 3,
    'groupAdvanced': 8,
    'groupWinner': 12,
    'semifinalist': 16,
    'finalist': 20,
    'winner': 25,
    'bestScorer': 25,
    'bestPlayer': 25,
    'totalExactGoals': 30,
    'totalGoals': 15
  }
}