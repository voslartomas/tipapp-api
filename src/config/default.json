{
  "host": "0.0.0.0",
  "port": 7300,
  "logging": {
    "level": "debug"
  },
  "origin": "http://localhost:3000",
  "swagger": {
    "enabled": true,
    "config_path": "./definitions/swagger.json"
  },
  "db": {
      "database": "tipapp",
      "username": "postgres",
      "password": "devpass",
      "host": "localhost",
      "dialect": "postgres",
      "port": 5436,
      "modelPath": "/../models/*.model.js"
  },
  "redis": {
    "host": "localhost",
    "port": 6379
  },
  "appcenter_push_token": "aa1a124fa04f708bc2e0dd4e6b60c5e7650a4e8d",
  "security": true,
  "points": {
    "hockey": {
      "matchWinner": 4,
      "exactScore": 8,
      "goalsOfOneTeam": 1,
      "goalsDifference": 2,
      "scorer": 8,
      "bestScorer": 2,
      "secondBestScorer": 3,
      "thirdBestScorer": 4,
      "fourthBestScorer": 5,
      "nhlSeries": 8,
      "nhlSeriesWithGames": 14,
      "nhlConference": 20,
      "nhlWinner": 40,
      "nhlMvp": 30,
      "wcGroup": 20,
      "wcLast": 20,
      "wcWinner": 40,
      "wcScorer": 30,
      "wcGoalkeeper": 30,
      "wcMvp": 30,
      "wcExactTotalGoals": 50,
      "wcTotalGoals": 20
    },
    "football": {
      "matchWinner": 3,
      "matchWinnerWithDifference": 4,
      "draw": 4,
      "exactScore": 6,
      "scorer": 5,
      "bestScorer": 3,
      "groupAdvanced": 8,
      "groupWinner": 12,
      "semifinalist": 16,
      "finalist": 20,
      "tournamentWinner": 25,
      "tournamentBestScorer": 25,
      "tournamentBestPlayer": 25,
      "totalExactGoals": 30,
      "totalGoals": 15
    }
  }
}
