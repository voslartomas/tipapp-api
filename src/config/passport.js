'use strict'
const bCrypt = require('bcrypt-nodejs')
const LocalStrategy = require('passport-local').Strategy

module.exports = (passport, userModel) => {
    
  passport.serializeUser((userModel, done) => done(null, userModel.id))

  passport.deserializeUser(async (id, done) => {
    const user = await userModel.findById(id)
    user ? done(null, user.get()) : done(user.errors, null)
  })

  passport.use('local-signup', new LocalStrategy(
    {
        usernameField: 'username',
        passwordField: 'password',
        email: 'email',
        passReqToCallback: true // allows us to pass back the entire request to the callback
    },

    async (req, username, password, done)  => {

      const generateHash = password => bCrypt.hashSync(password, bCrypt.genSaltSync(8), null)

        const userNick = await userModel.findOne({
          where: { username: username }
        })

        const userEmail = await userModel.findOne({
          where: { email: req.body.email }
        })

        if (userNick || userEmail) {
          if(userNick){
            return done(null, false, { message: 'Username is already taken' })
          } else {
            return done(null, false, { message: 'Email address is already taken' })
          }
        } else {
          const userPassword = generateHash(password)
          const data = { username: username,
            password: userPassword,
            firstName: req.body.firstName,
            lastName: req.body.lastName,
            email: req.body.email.toLowerCase()
          }

          await userModel.create(data).then((newUser, created) => {
            newUser ? done(null, newUser) : done(null, false)
          })
        }
    }
    ))

  passport.use('local-signin', new LocalStrategy(
    {
      usernameField: 'username',
      passwordField: 'password',
      passReqToCallback: true // allows us to pass back the entire request to the callback
    },

    async (req, username, password, done) => {
      try {
        const isValidPassword = (userpass, password) => bCrypt.compareSync(password, userpass)
        const user = await userModel.findOne({
          where: {
            username: username
          }
        })
        if (!user) {
          return done(null, false, { message: 'Username does not exist' })
        }
        if (!isValidPassword(user.password, password)) {
          return done(null, false, { message: 'Incorrect password.' })
        }
        const userInfo = user.get()
          return done(null, userInfo)
        } catch (err) {
          console.log("Error:", err)
          return done(null, false, { message: 'Something went wrong with your Signin' })
        }
    }
  ))
}