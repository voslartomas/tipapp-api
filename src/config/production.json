{
  "port": 8080,
  "logging": {
    "level": "info"
  },
  "origin": "*",
  "swagger": {
    "enabled": false
  },
  "db": {
    "database": "ywucuvfk",
    "username": "ywucuvfk",
    "password": "w6Ew55-nQHwKNI3VRZ0Ovsd0ySIYQZm8",
    "host": "tyke.db.elephantsql.com",
    "dialect": "postgres",
    "dialectOptions": {
      "ssl": true
    },
    "logging": true,
    "port": 5432,
    "modelPath": "/../models/*.model.js"
  },
  "redis": {
    "host": "localhost",
    "port": 6379
  },
  "security": true
}
