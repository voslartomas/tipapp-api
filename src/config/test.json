{
  "logging": {
    "level": "error"
  },
  "port": 7301,
  "swagger": {
    "config_path": "./src/definitions/swagger.json"
  },
  "db": {
      "storage": "test-db.sqlite",
      "databaseName": "test-db.sqlite",
      "modelPath": "/../models/*.model.ts"
  },
  "security": false
}
