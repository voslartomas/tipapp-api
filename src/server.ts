import * as config from 'config'
import * as express from 'express'
import { createLogger } from 'bunyan'
import * as http from 'http'
import { Inject } from 'typescript-ioc'
import AppLogger from './services/logger'
import { Server as RestServer } from 'typescript-rest'
import routes from './routes'
import { Sequelize } from 'sequelize-typescript'
import loggerMiddleware from './middlewares/logger'
import Database from './services/database'
import JWTPassport from './security/passport'
import PushNotifications from './services/pushNotifications'
import * as cors from 'cors'
import * as cron from 'node-cron'
import { AddressInfo } from 'net'

export class Server {

  private app: express.Application
  private server: http.Server = undefined
  @Inject
  private logger: AppLogger

  @Inject
  private database: Database

  @Inject
  private jwtPassport: JWTPassport

  @Inject
  private pushNotifications: PushNotifications

  constructor() {
    this.app = express()
    this.logger = undefined
    this.config()

    this.app.use(function(req, res, next) {
      res.header('Access-Control-Allow-Origin', '*')
      res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization')
      next()
    })

    if (config.get('security')) {
      this.app.get('*', (req, res, next) => {
        this.jwtPassport.getPassport().authenticate('jwt', { session: false })(req, res, next)
      })

      this.app.post('*', (req, res, next) => {
        if (req.originalUrl === '/login' || req.originalUrl === '/api/register') {
          return next()
        }

        this.jwtPassport.getPassport().authenticate('jwt', { session: false })(req, res, next)
      })

      this.app.put('*', (req, res, next) => {
        this.jwtPassport.getPassport().authenticate('jwt', { session: false })(req, res, next)
      })

      this.app.delete('*', (req, res, next) => {
        this.jwtPassport.getPassport().authenticate('jwt', { session: false })(req, res, next)
      })
    }

    RestServer.buildServices(this.app, ...routes)

    if (config.get('swagger.enabled')) {
      RestServer.swagger(this.app, { filePath: `${config.get('swagger.config_path')}`, endpoint: '/api-docs', host: `localhost:${config.get('port')}`, schemes: ['http'] })
    }
  }

  public getApp() {
    return this.app
  }

  /**
   * Configure the express app.
   */
  private config(): void {
    this.app.use(loggerMiddleware(this.logger))
    this.app.use(this.jwtPassport.getPassport().initialize())

    const options: cors.CorsOptions = {
      allowedHeaders: ['Origin', 'X-Requested-With', 'Content-Type', 'Accept', 'Authorization'],
      credentials: true,
      methods: 'GET,HEAD,OPTIONS,PUT,PATCH,POST,DELETE',
      origin: config.get('origin'),
      preflightContinue: false
    }

    this.app.use(cors(options))

    this.app.options('*', cors(options))
  }

  /**
   * Start the server
   * @returns {Promise<any>}
   */
  public start(): Promise<any> {
    cron.schedule('* * * * *', () => {
      console.log('Running push notification service')
      this.pushNotifications.sendPushNotifications()
    })

    return new Promise<any>((resolve, reject) => {
      this.server = this.app.listen(config.get('port'), config.get('port'), () => {
        const { port } = this.server.address() as AddressInfo
        this.logger.info(`Listening to http://localhost:${port}`)
        return resolve(true)
      }).on('error', (err) => reject(err))
    })
  }

  /**
   * Stop the server (if running).
   * @returns {Promise<boolean>}
   */
  public stop(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
      if (this.server) {
        this.server.close(() => {
          return resolve(true)
        })
      } else {
        return resolve(true)
      }
    })
  }
}
