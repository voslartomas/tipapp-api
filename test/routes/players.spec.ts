import * as config from 'config'
import { Server } from '../../src/server'
import * as request from 'supertest'
import { Server as RestServer, HttpMethod } from 'typescript-rest'

jest.mock('../../src/services/database')

// let app = undefined
const app = new Server().getApp()
describe('Player API', () => {

  it('GET /players should return list of players', async () => {
    const response = await request(app).get('/api/players')
    expect(response.status).toBe(200)
    expect(response.body).toMatchSnapshot()
  })

  it('GET /players/1 should return 200 when user is in db', async () => {
    const response = await request(app).get('/api/players/1')
    expect(response.status).toBe(200)
    expect(response.body).toMatchSnapshot()
  })

  it('GET /players/2 should return 404 when user not found in db', async () => {
    const response = await request(app).get('/api/players/2')
    expect(response.status).toBe(404)
  })

  it('POST /players should return 200', async () => {
    const response = await request(app).post('/api/players').send({ name: 'Jaromir Jar' })
    expect(response.status).toBe(200)
    expect(response.body).toMatchSnapshot()
  })

  it('PUT /players should return 204', async () => {
    const response = await request(app).put('/api/players/1').send({ name: 'Jaromir Jagr' })
    expect(response.status).toBe(204)
    expect(response.body).toMatchSnapshot()
  })

  it('PUT /players should return 200 and creates new user if does not exist already', async () => {
    const response = await request(app).put('/api/players/2').send({ username: 'Sidney Crosby' })
    expect(response.status).toBe(200)
    expect(response.body).toMatchSnapshot()
  })

  it('DELETE /players should return 204 user found', async () => {
    const response = await request(app).delete('/api/players/1')
    expect(response.status).toBe(204)
  })
})
