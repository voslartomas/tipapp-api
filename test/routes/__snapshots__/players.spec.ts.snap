// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player API GET /players should return list of players 1`] = `
Array [
  Object {
    "testdata": "shared",
  },
]
`;

exports[`Player API GET /players/1 should return 200 when user is in db 1`] = `
Object {
  "testData": "shared",
}
`;

exports[`Player API POST /players should return 200 1`] = `
Object {
  "name": "Jaromir Jar",
}
`;

exports[`Player API PUT /players should return 200 and creates new user if does not exist already 1`] = `
Object {
  "username": "Sidney Crosby",
}
`;

exports[`Player API PUT /players should return 204 1`] = `Object {}`;
